### PR TITLE
Change add place button dynamically

### PR DESCRIPTION
**Pull request by zhoua0805**
This PR fixes some small issues of the itinerary page

**1.** Create function renderPlaceButtons(). 
  - **Context**: The add button of each place needs to be enabled/disabled after user adds a place/ deletes an event that used to be a place/ changes to a new list to work on. 
- Previously, the renderPlaces function is called every time these events happen, which clears the places container and fetches all the place details from the google maps api again. Therefore, a lot of redundant requests are made.
- **Fix**: The new function only changes the button, and keeps everything else the same
- if the parameter placeId is null, the function will go through every place and update their buttons, else, it will only toggle the provided place's button

**2.** Remove the line that calls the sigup.css file in itinerary.html since it's affecting the page layout

**3.** Update starting address in sessionStorage
 - **Context**: The starting address is stored in sessionStorage and is updated using an input onChange function. However, when we use the google maps autocomplete feature to fill out the address, the onChange function skips it (for example, if we type in "to" and google maps autocompletes to "toronto", the value stored in sessionStorage will still be "to"
- **Fix**: add an event listener to the autocomplete to update sessionStorage.

### CODE REVIEW

File: planet/src/main/webapp/js/placesItinerary.js [hunk 1]
```diff
@@ -236,3 +245,32 @@ function validatePlaceDuration(duration) {
     }
     return true;
 }
+
+export function renderPlaceButtons(placeId, enableAdd) {
+    if (placeId) {
+        togglePlaceButton(placeId, enableAdd);
+    } else {
+        const userId = firebase.auth().currentUser.uid;
```
Comment (nabil-dbz) [inline planet/src/main/webapp/js/placesItinerary.js]: It's safer to check that the currentUser is different than null before trying to access the uid property.

Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: I can quickly add it in, but there won't be any place buttons if the user is not signed in, so it's almost impossible for this edge case to happen

Comment (azhang72) [reply planet/src/main/webapp/js/placesItinerary.js]: Is it possible for the call to firebase.auth().currentUser.uid to fail? Trying to think of any scenarios where |userId| could be null here..

File: planet/src/main/webapp/js/placesItinerary.js [hunk 2]
```diff
@@ -236,3 +245,32 @@ function validatePlaceDuration(duration) {
     }
     return true;
 }
+
+export function renderPlaceButtons(placeId, enableAdd) {
```
Comment (EkkiD) [inline planet/src/main/webapp/js/placesItinerary.js]: I don't follow what the second parameter here does. What is the different behaviour if it's true or false?

File: planet/src/main/webapp/js/placesItinerary.js [hunk 3]
```diff
@@ -236,3 +245,32 @@ function validatePlaceDuration(duration) {
...
+        });
+    }
+} 
+
+function togglePlaceButton(placeId, enableAdd) {
+    const placeButton = document.getElementById('place-button-' + placeId);
+    if (placeButton) {
+        if (enableAdd) {
+            placeButton.setAttribute('onclick', 'openAddPlaceForm(event, "' + placeId + '")');
+            placeButton.innerHTML = `<i class="material-icons small">playlist_add</i>`;
```
Comment (EkkiD) [inline planet/src/main/webapp/js/placesItinerary.js]: all of the actual logic of this function is different depending on the enableAdd parameter. It seems it it would be clearer as two functions with descriptive names

Reference:
https://docs.google.com/document/d/1R5iFoLp41lbIrHvOpViblqH91j-VscdkXNNAL6uYxlU/edit

File: planet/src/main/webapp/js/placesItinerary.js [hunk 4]
```diff
@@ -236,3 +245,32 @@ function validatePlaceDuration(duration) {
...
+    if (placeId) {
+        togglePlaceButton(placeId, enableAdd);
+    } else {
+        const userId = firebase.auth().currentUser.uid;
+        const placeListRef = database.ref('users/' + userId + '/places');
+        placeListRef.once('value', (placeListSnapshot) => {
+            placeListSnapshot.forEach(function(childPlace) {
+                let placeId = childPlace.key;
+                const noSuchEvent = document.getElementById(placeId) === null;
+                togglePlaceButton (placeId, noSuchEvent);
```
Comment (EkkiD) [inline planet/src/main/webapp/js/placesItinerary.js]: use positive booleans: https://docs.google.com/document/d/1ljFkOXo6NROsfKGtASDPRTPUq0oDPpm8jc2D5hBfqLE/preview

File: planet/src/main/webapp/js/placesItinerary.js [hunk 5]
```diff
@@ -38,10 +38,18 @@ function initAutocomplete() {
...
+    autocompleteStart = new google.maps.places.Autocomplete(startAddress,options);
+    autocompleteStart.setFields(['address_component']);
+    autocompleteStart.addListener('place_changed', updateStartingAddress);
+
     autocompleteEvent = new google.maps.places.Autocomplete(eventAddress,options);
 }
 
+function updateStartingAddress() {
+    const startAddress = document.getElementById('starting-address');
+    sessionStorage.setItem('start', startAddress.value);
```
Comment (EkkiD) [inline planet/src/main/webapp/js/placesItinerary.js:50]: is this related to this PR? It looks like a separate fix

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/placesItinerary.js (hunk 5 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js:50]: I talked about it in the PR description (since all 3 fixes are relatively small I combined them into one PR)

File: planet/src/main/webapp/js/placesItinerary.js (hunk 2 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: If a place haven't been added, the button should be enabled, and it will be disabled as soon as the user adds the place.
Meanwhile, if the user deletes an event that was added from a place, the button should be enabled again.

> Example:

      user has a place called CN Tower --> user adds CN Tower as an event --> button disabled
                                       --> user deletes the CN Tower event --> button enabled

File: planet/src/main/webapp/js/placesItinerary.js (hunk 1 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: hmm not sure, if the user is signed in I can't think of any other scenarios that the userid would be null

File: planet/src/main/webapp/js/placesItinerary.js (hunk 3 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: Ah okay, do you mean I should change the parameter name of enableAdd? If so, what should I change it to?
Or do you mean I should change the function to two functions such as enablePlaceButton() and disablePlaceButton()

File: planet/src/main/webapp/js/placesItinerary.js (hunk 4 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: Okay!

File: planet/src/main/webapp/js/placesItinerary.js (hunk 1 above)
Comment (nabil-dbz) [reply planet/src/main/webapp/js/placesItinerary.js]: If in you're sure that the function is only called when the user is signed in, then it should be good as it is. I usually take precautions and verify that currentUser is not null to avoid errors popping up in the console.

**Review by nabil-dbz — APPROVED** [review]

File: planet/src/main/webapp/js/placesItinerary.js (hunk 1 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: Okay! thanks!

File: planet/src/main/webapp/js/placesItinerary.js [hunk 6]
```diff
@@ -38,10 +38,18 @@ function initAutocomplete() {
     let options = {
         types: ['geocode']
     };
-    autocompleteStart = new google.maps.places.Autocomplete(startAddress,options); 
+    autocompleteStart = new google.maps.places.Autocomplete(startAddress,options);
+    autocompleteStart.setFields(['address_component']);
+    autocompleteStart.addListener('place_changed', updateStartingAddress);
+
     autocompleteEvent = new google.maps.places.Autocomplete(eventAddress,options);
```
Comment (azhang72) [inline planet/src/main/webapp/js/placesItinerary.js:45]: Will you need similar listeners and fields for |autocompleteEvent| as well?

File: planet/src/main/webapp/js/placesItinerary.js (hunk 1 above)
Comment (azhang72) [reply planet/src/main/webapp/js/placesItinerary.js]: What about network failure?

Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: true, i'll add it in then! thank you!

File: planet/src/main/webapp/js/placesItinerary.js (hunk 6 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js:45]: When I tested it it seems like the input values are automatically changed with autocomplete. I only need the listener for autocompleteStart since we need to store the start location in sessionStorage

File: planet/src/main/webapp/js/placesItinerary.js (hunk 3 above)
Comment (EkkiD) [reply planet/src/main/webapp/js/placesItinerary.js]: > Or do you mean I should change the function to two functions such as enablePlaceButton() and disablePlaceButton()

This is what I meant

File: planet/src/main/webapp/js/placesItinerary.js (hunk 2 above)
Comment (EkkiD) [reply planet/src/main/webapp/js/placesItinerary.js]: It looks like we're putting several different functions into one here and controlling the code path with parameters. That makes it very difficult to read.

Q: What does renderPlaceButtons do?
A: Well, it renders all the places that the user has saved, except when it's called with a place id, in which case it only renders that place. It accepts an enableAdd boolean, but it only uses it if placeId is also set.

A function should do exactly one thing, which should be described by its name. Specifically to this case, this should probably be 3 functions:
disablePlaceButton(placeId)
enablePlaceButton(placeId)
renderPlaceButtons()

**Review by EkkiD — CHANGES_REQUESTED** [review]

File: planet/src/main/webapp/js/placesItinerary.js (hunk 2 above)
Comment (zhoua0805) [reply planet/src/main/webapp/js/placesItinerary.js]: That makes sense! Thank you :D

**Review by EkkiD — APPROVED** [review]
LGTM
